// react-watermark-component/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _WaterMarkCanvas = require('./WaterMarkCanvas');

var _WaterMarkCanvas2 = _interopRequireDefault(_WaterMarkCanvas);

var _SecurityDefense = require('./SecurityDefense');

var _SecurityDefense2 = _interopRequireDefault(_SecurityDefense);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var defaultOptions = {
  chunkWidth: 200,
  chunkHeight: 60,
  textAlign: 'left',
  textBaseline: 'bottom',
  globalAlpha: 0.17,
  font: '14px Microsoft Yahei',
  rotateAngle: -0.26,
  fillStyle: '#666'
};

var waterMarkStyle = 'position: absolute;left: 0;right: 0;top:0;bottom:0;opacity: 0.7;z-index: 9999;pointer-events: none;overflow: hidden;background-color: transparent;background-repeat: repeat;';
var noop = function noop() {};

var WaterMark = (_temp2 = _class = function (_React$Component) {
  _inherits(WaterMark, _React$Component);

  function WaterMark() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, WaterMark);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = WaterMark.__proto__ || Object.getPrototypeOf(WaterMark)).call.apply(_ref, [this].concat(args))), _this), _this.getStyles = function () {
      var _this$props = _this.props,
          waterMarkText = _this$props.waterMarkText,
          options = _this$props.options;

      var newOptions = Object.assign({}, defaultOptions, options);
      return (0, _WaterMarkCanvas2.default)(waterMarkText, newOptions);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(WaterMark, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var dom = document.getElementById('water-mark-observer');
      var parent = document.getElementById('water-mark-wrapper');
      var securityAlarm = this.props.securityAlarm;

      var security = new _SecurityDefense2.default(waterMarkStyle, this.getStyles, securityAlarm);
      security.registerNodeRemoveListener(dom, parent);
      security.registerNodeAttrChangeListener(dom);
    }
  }, {
    key: 'render',
    value: function render() {
      var children = this.props.children;

      var styles = {
        position: 'absolute',
        left: 0,
        right: 0,
        top: 0,
        bottom: 0,
        opacity: 0.7,
        zIndex: 9999,
        pointerEvents: 'none',
        overflow: 'hidden',
        backgroundImage: 'url("' + this.getStyles() + '")',
        backgroundColor: 'transparent',
        backgroundRepeat: 'repeat'
      };

      return _react2.default.createElement(
        'div',
        { style: { position: 'relative' }, id: 'water-mark-wrapper' },
        _react2.default.createElement('div', { style: styles, id: 'water-mark-observer' }),
        children
      );
    }
  }]);

  return WaterMark;
}(_react2.default.Component), _class.propTypes = {
  children: _propTypes2.default.element.isRequired,
  waterMarkText: _propTypes2.default.string.isRequired,
  securityAlarm: _propTypes2.default.func,
  options: _propTypes2.default.shape({
    chunkWidth: _propTypes2.default.number,
    chunkHeight: _propTypes2.default.number,
    textAlign: _propTypes2.default.string,
    textBaseline: _propTypes2.default.string,
    globalAlpha: _propTypes2.default.number,
    font: _propTypes2.default.string,
    rotateAngle: _propTypes2.default.number,
    fillStyle: _propTypes2.default.string
  })
}, _temp2);


WaterMark.defaultProps = {
  securityAlarm: noop,
  options: defaultOptions
};

exports.default = WaterMark;
module.exports = exports['default'];
