// react-watermark-component/dist/SecurityDefense.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var SecurityDefense = function SecurityDefense(waterMarkStyle, getImageUrl, securityAlarm) {
  var _this = this;

  _classCallCheck(this, SecurityDefense);

  this.getWaterMarkDom = function () {
    return document.getElementById(_this.randomId);
  };

  this.registerNodeRemoveListener = function (dom, parent) {
    dom.addEventListener('DOMNodeRemoved', function () {
      _this.createWaterMarkDom(parent);
      if (_this.securityAlarm) {
        _this.securityAlarm();
      }
    }, false);
  };

  this.createWaterMarkDom = function (parent) {
    var newWaterMark = document.createElement('div');
    _this.randomId = 'water-mark-observer-' + new Date().getTime();
    newWaterMark.id = _this.randomId;
    _this.waterMarkStyle = _this.waterMarkStyle.concat('background-image: url("' + _this.getImageUrl() + '");');
    newWaterMark.style = _this.waterMarkStyle;
    parent.appendChild(newWaterMark);
    var newDom = _this.getWaterMarkDom();
    _this.registerNodeRemoveListener(newDom, parent);
    _this.registerNodeAttrChangeListener(newDom);
  };

  this.registerNodeAttrChangeListener = function (target) {
    var MutationObserver = window.MutationObserver || window.WebKitMutationObserver || window.MozMutationObserver;
    var observer = new MutationObserver(function (mutations) {
      mutations.forEach(function (mutation) {
        if (mutation.type === 'attributes') {
          target.parentNode.removeChild(target);
          observer.disconnect();
        }
      });
    });
    var config = { attributes: true, attributeFilter: ['style'] };
    observer.observe(target, config);
  };

  this.randomId = 'water-mark-observer';
  this.waterMarkStyle = waterMarkStyle;
  this.getImageUrl = getImageUrl;
  this.securityAlarm = securityAlarm;
};

exports.default = SecurityDefense;
module.exports = exports['default'];
